### PR TITLE
[DF] Add convenience function to describe dataset

### PR DESCRIPTION
**Pull request by stwunsch**
See the linked issue (#7561) for the discussion of the programming model

### CODE REVIEW

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 1]
```diff
@@ -1873,6 +1873,51 @@ public:
                                                     convertVector2RVec);
    }
 
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Return the schema of the dataset
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:1877]:
```suggestion
   /// \brief Return the schema of the dataset.
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 2]
```diff
@@ -1873,6 +1873,51 @@ public:
...
+   /// \return the schema of the dataset as string
+   ///
+   /// This convenience function combines the information from GetColumnNames and
+   /// GetColumnType in a table to inspect easily the dataset schema.
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:1881]:
```suggestion
   /// This convenience function combines the information from GetColumnNames() and
   /// GetColumnType() in a table to inspect easily the dataset schema.
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 3]
```diff
@@ -1873,6 +1873,51 @@ public:
...
+      const auto minColumnSpace = 8u; // Minimum space between columns
+      columnWidth = (columnWidth / minColumnSpace + 1) * minColumnSpace;
+
+      // Build table of column names and types in a string
+      std::stringstream ss;
+      for (auto i = 0u; i < n; i++) {
+         ss << std::left << std::setw(columnWidth) << columnNames[i]
+            << columnTypes[i];
+         if (i < n - 1)
+             ss << "\n";
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:1915]:
```suggestion
             ss << '\n';
```

**Review by eguiraud — COMMENTED** [review]
Nice!! I think we should discuss the points you raise in #7561 before merging, but this is basically it.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 4]
```diff
@@ -1873,6 +1873,51 @@ public:
...
+      const auto columnTypes = GetColumnTypeNamesList(columnNames);
+
+      // Compute optimal width for the column names
+      const auto n = columnNames.size();
+      auto columnWidth = 0u;
+      for (const auto& name : columnNames) {
+         const auto length = name.length();
+         if (length > columnWidth)
+             columnWidth = length;
+      }
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:1905]:
```cpp
const auto columnWidth = std::max_element(columnNames.begin(), columnNames().end(), [] (const std::string &s) { return s.length(); })->length();
```

or maybe not...damn STL algorithms are unreadable

Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:1905]: That's why I stick to good old C style coding if it's not magic ;)

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 5]
```diff
@@ -1999,6 +2000,100 @@ public:
                                                     convertVector2RVec);
    }
 
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Return information about the dataframe.
+   /// \return information about the dataframe as string
+   ///
+   /// This convenience function describes the dataframe and combines following information:
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2007]:
```suggestion
   /// This convenience function describes the dataframe and combines the following information:
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 6]
```diff
@@ -1999,6 +2000,100 @@ public:
...
+   /// - Number of event loops run, see GetNRuns()
+   /// - Number of total and defined columns, see GetColumnNames() and GetDefinedColumnNames()
+   /// - Name of the columns, see GetColumnNames()
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2010]:
```suggestion
   /// - Column names, see GetColumnNames()
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 7]
```diff
@@ -1999,6 +2000,100 @@ public:
...
+   /// - Type of the columns, see GetColumnType()
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2011]:
```suggestion
   /// - Column types, see GetColumnType()
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 8]
```diff
@@ -1999,6 +2000,100 @@ public:
...
+   ///
+   /// Please note that this is a convenience feature and the layout of the output can be subject
+   /// to change and should not be automatically parsed.
+   ///
+   /// ### Example usage:
+   /// ~~~{.cpp}
+   /// RDataFrame df(10);
+   /// auto df2 = df.Define("x", "1.f").Define("s", "\"myStr\"");
+   /// // Describe the dataframe
+   /// std::cout << df2.Desribe() << std::endl;
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2024]:
```suggestion
   /// std::cout << df2.Describe() << std::endl;
```

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 9]
```diff
@@ -1999,6 +2000,100 @@ public:
...
+   /// ~~~
+   ///
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2026]: should mention differences between calling `Describe` on the head node or on other nodes, following GetDefinedColumnNames or others

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 10]
```diff
@@ -1999,6 +2000,100 @@ public:
...
+          "Columns in total",
+          "Columns from defines",
+          "Event loops run",
+          "Processing slots"};
+      const std::vector<std::string> metadataValues = {
+          std::to_string(columnNames.size()),
+          std::to_string(definedColumnNamesSet.size()),
+          std::to_string(GetNRuns()),
+          std::to_string(GetNSlots())
+          };
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2047]: formatting looks a bit funny, might need a clang-format pass?

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx [hunk 11]
```diff
@@ -1999,6 +2000,100 @@ public:
...
+      // The column width of the values is required to make right-bound numbers and is equal
+      // to the maximum of the string "Value" and all values to be put in this column.
+      const auto columnWidthValues = std::max(
+              std::max_element(metadataValues.begin(), metadataValues.end())->size(),
+              static_cast<std::size_t>(5u));
+      std::stringstream ss;
+      ss << std::left << std::setw(columnWidthProperties) << "Property"
+         << std::left << std::setw(columnWidthValues) << "Value\n"
+         << std::left << std::setw(columnWidthProperties) << "--------"
+         << std::left << std::setw(columnWidthValues) << "-----\n";
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2060]: isn't `std::left` the default? also if I'm not mistaken setting it once is enough

**Review by eguiraud — CHANGES_REQUESTED** [review]
Thanks a lot @stwunsch . See some minor comments inline, plus please add a mention of `Describe` to the cheat-sheet in RDataFrame.cxx

**Comment by stwunsch** [comment]
@phsft-bot build

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 10 above)
Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2047]: Good point, haven't run clang-format at all 😬 I'll do this after addressing the other comments.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 11 above)
Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2060]: Yep it is! I'll remove it in various places, also below.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 9 above)
Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2026]: You mean that it returns just the state up to the selected node? Or is there any other detail I've missed?

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 10 above)
Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx]: Clang-format does weird stuff :/ /I'll keep it like this now.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 9 above)
Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2028]: See the sentence added above.

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 11 above)
Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx]: Done!

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 10 above)
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx]: what do you mean? we try to format everything with clang-format in RDF. for very weird behavior you can enclose blocks of code in `// clang-format off/on` pairs (we do so for doxygen docs, for example).

Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx]: alright, clang-format applied :) i find it not really readable, but i'm happy to following the guidlines ;)

**Comment by stwunsch** [comment]
@eguiraud rebased and integrated the requested changes

File: tree/dataframe/inc/ROOT/RDF/RInterface.hxx (hunk 9 above)
Comment (stwunsch) [reply tree/dataframe/inc/ROOT/RDF/RInterface.hxx:2028]: Last comment standing!

**Comment by eguiraud** [comment]
Hey, this looks great, please add a mention of Describe to the cheat-sheet in RDataFrame.cxx and it's good to go

**Comment by stwunsch** [comment]
> Hey, this looks great, please add a mention of Describe to the cheat-sheet in RDataFrame.cxx and it's good to go

@eguiraud Done!

**Comment by eguiraud** [comment]
Sorry _last thing_, I promise: can you add a line in the 6.26 release notes about this? So we don't depend on me remembering :smile:

**Comment by stwunsch** [comment]
> Sorry _last thing_, I promise: can you add a line in the 6.26 release notes about this please? So we don't depend on me remembering 😄

@eguiraud Also done! :D You may want to rephrase the release notes at some point, but I leave this for the future you ;)

**Review by eguiraud — APPROVED** [review]
Thanks a lot Stefan, super nice work

**Pull request merged**